manager/src/main/scala/org/broadinstitute/transporter/queue/QueueController.scala

```diff
@@ -0,0 +1,112 @@
+package org.broadinstitute.transporter.queue
+
+import cats.effect.{ExitCase, IO}
+import cats.implicits._
+import io.chrisdavenport.fuuid.FUUID
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import org.broadinstitute.transporter.db.DbClient
+import org.broadinstitute.transporter.kafka.KafkaClient
+
+/**
+  * Component responsible for handling all queue-related requests.
+  *
+  * @param dbClient client which can interact with Transporter's DB
+  * @param kafkaClient client which can interact with Transporter's Kafka cluster
+  */
+class QueueController(dbClient: DbClient, kafkaClient: KafkaClient) {
+  import QueueController._
+
+  private val logger = Slf4jLogger.getLogger[IO]
+
+  /**
+    * Create a new queue matching the parameters of the given request,
+    * returning all data for the new resource on success.
+    *
+    * Fails with a [[QueueAlreadyExists]] if trying to create a queue
+    * with a name that's already registered in the DB.
+    */
+  def createQueue(request: QueueRequest): IO[Queue] =
+    for {
+      preexisting <- dbClient.lookupQueue(request.name)
+      _ <- preexisting.fold(IO.unit) { _ =>
+        IO.raiseError(QueueAlreadyExists(request.name))
+      }
+      queue <- initializeQueue(request)
+    } yield {
+      queue
+    }
+
+  /**
+    * Initialize resources for a new queue matching the parameters
+    * of the given request, returning all data for the new resource
+    * on success.
+    *
+    * The process of initializing a queue requires making updates to
+    * both the DB and Kafka. To avoid inconsistent state, we:
+    *   1. Enter new queue information into the DB
+    *   2. Attempt to create Kafka topics matching the new DB info
```

Comment:
Isn't it still possible to have dangling queues in the db if something catastrophic happens between #2 and #3?  Do we care?  If we do what are the plans for cleaning these up?

Reply:
It is possible... I think there's a way to force the DB transaction to stay open while creating the queues, so that if it blows up in the middle the transaction gets rolled back. I'll investigate.